src/sentry/grouping/strategies/hierarchical.py

```diff
@@ -176,16 +176,24 @@ def _accumulate_frame_levels(values, blaming_frame_idx, depth, direction):
     # subtract depth by one to count blaming frame
     depth -= 1
 
-    idx = blaming_frame_idx + direction
-    while 0 <= idx < len(values) and depth > 0:
-        component = values[idx]
-        idx += direction
+    start = blaming_frame_idx + direction
+    if start < 0:
+        return rv
 
+    added = 0
+    prev_was_prefix = values[blaming_frame_idx].is_prefix_frame
+    for component in values[start:None:direction]:
         if not component.contributes:
             continue
 
-        rv.append(component)
-        if not component.is_prefix_frame:
-            depth -= 1
+        if prev_was_prefix:
+            rv.append(component)
+        else:
+            if added == depth:
+                break
+            rv.append(component)
+            added += 1
```

Comment:
```suggestion
        if added == depth:
            break

        rv.append(component)

        if not prev_was_prefix:
            added += 1
```

idea: if we exhaust our depth and then hit a run of non-contributing frames, we iterate for nothing, so we should check depth one time too often to break asap

so actually `if added == depth` should be sorted before `if not component.contributes`

Reply:
break asap is a good idea, but In case of `depth=0` we want to reach `rv.append(component)` in case the blaming frame is a prefix, right?
We could also achieve this before the loop though.

Reply:
ah yeah

ok I think this is fine, I think my suggestion may have regressed something anyway